### PR TITLE
added shared pointer to matrix. changed c++ standard to 17

### PR DESCRIPTION
**Pull request by shirnschall**
## Summary
changed data pointer to shared pointer. We switched to c++ standard 17 for the shared_ptr to call delete[] instead of delete.

## Quick changelog

- c++ standard changed to 17 from 11
- changed T* data_ to std::shared_ptr<T[]> data_

### CODE REVIEW

File: src/matrix.h [hunk 1]
```diff
@@ -45,13 +45,7 @@ namespace bla
...
         }
 
-        ~Matrix() 
-        { 
-            if(DelData())
-            {
-                delete[] data_;
-            } 
-        }
+        ~Matrix() { ; }
```
Comment (erdieee) [inline src/matrix.h:48]: We should be consistent with empty brackets. Use {} as in the constructor instead of { ; }.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,7 +1,7 @@
 cmake_minimum_required(VERSION 3.20)
 project(bla)
 
-set (CMAKE_CXX_STANDARD 11)
+set (CMAKE_CXX_STANDARD 17)
```
Comment (erdieee) [inline CMakeLists.txt:4]: Could we also use 20?

**Comment by erdieee** [comment]
LGTM 👍

File: CMakeLists.txt (hunk 2 above)
Comment (shirnschall) [reply CMakeLists.txt:4]: Let's stick to 17 as we do not need c++ 20 or 23 features at the moment. This way the project stays compatible with a wider range of compilers, making it more accessible.

**Pull request merged**